blocks/log/log.py

```diff
@@ -0,0 +1,136 @@
+"""The event-based main loop of Blocks."""
+import sqlite3
+from abc import ABCMeta
+from collections import defaultdict
+from numbers import Integral
+from uuid import uuid4
+
+import six
+
+
+@six.add_metaclass(ABCMeta)
+class _TrainingLog(object):
+    """Base class for training log.
+
+    A training log stores the training timeline, statistics and other
+    auxiliary information. Training logs can use different backends e.g.
+    in-memory Python objects or an SQLite database.
+
+    Information is stored similar to a nested dictionary, so use
+    ``log[time][key]`` to read data. An entry without stored data will
+    return an empty dictionary-like object that can be written to,
+    ``log[time][key] = value``.
+
+    Depending on the backend, ``log[time] = {'key': 'value'}`` could fail.
+    Use ``log[time].update({'key': 'value'})`` for compatibility across
+    backends.
+
+    In addition to the set of records displaying training dynamics, a
+    training log has a :attr:`status` attribute, which is a dictionary with
+    data that is not bound to a particular time.
+
+    .. warning::
+
+       Changes to mutable objects might not be reflected in the log,
+       depending on the backend. So don't use
+       ``log.status['key'].append(...)``, use ``log.status['key'] = ...``
```

Comment:
Gross. I wish there was some syntactic sugar we could pull to make this less error-prone.